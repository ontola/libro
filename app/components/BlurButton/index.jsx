import PropTypes from 'prop-types';
import React from 'react';

const propTypes = {
  /** The children float to the top right */
  children: PropTypes.node,
  /** The header floats to the top left */
  onClick: PropTypes.func,
};

// Used to remove the annoying focus outline border after clicking
const onClickAndBlur = (e, onClick) => {
  if (onClick !== undefined) {
    onClick(e);
    // Only blur when the event is a click, not for using the Enter button.
    // React fires a clickEvent when the user uses Enter, but the coordinates are zero.
  }
  if (e.nativeEvent.x !== 0) {
    e.currentTarget.blur();
  }
};

const BlurButton = ({
  children,
  onClick,
  ...props
}) => (
  <button
    {...props}
    onClick={(e) => onClickAndBlur(e, onClick)}
  >
    {children}
  </button>
);

BlurButton.propTypes = propTypes;

export default BlurButton;
