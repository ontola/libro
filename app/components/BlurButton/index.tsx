import { ButtonBase } from '@mui/material';
import React, {
  CSSProperties,
  MouseEvent,
} from 'react';

type OnClickHandler = (e: MouseEvent<HTMLButtonElement>) => any;

const preventAfterClickOutline = (e: MouseEvent<HTMLButtonElement>, onClick: OnClickHandler | undefined) => {
  const isClickByEnterKey = e.nativeEvent.x === 0;

  if (typeof onClick !== 'undefined') {
    onClick(e);
  }

  if (!isClickByEnterKey) {
    e.currentTarget.blur();
  }
};

export interface BlurButtonProps {
  className?: string;
  style?: CSSProperties;
  title?: string;
  onClick?: OnClickHandler;
}

const BlurButton = React.forwardRef<HTMLButtonElement, React.PropsWithChildren<BlurButtonProps>>(({
  children,
  onClick,
  ...props
}, ref) => (
  <ButtonBase
    ref={ref}
    type="button"
    {...props}
    onClick={(e) => preventAfterClickOutline(e, onClick)}
  >
    {children}
  </ButtonBase>
));

BlurButton.displayName = 'BlurButton';

export default BlurButton;
