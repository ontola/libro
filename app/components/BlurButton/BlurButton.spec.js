import BlurButton from './index';

const clickHandler = jest.fn();

argUnit(BlurButton, () => {
  it('should render', () => {
    expect(clickHandler).not.toHaveBeenCalled();
    expect(subject).toMatchSnapshot();
  });

  it('should blur when clicked', () => {
    const blur = jest.fn();
    const eventMock = {
      currentTarget: { blur },
      nativeEvent: { x: 1 },
    };
    expect(subject.simulate('click', eventMock));
    expect(clickHandler).not.toHaveBeenCalled();
    expect(blur).toHaveBeenCalled();
  });

  it('should not blur when press enter', () => {
    const blur = jest.fn();
    const eventMock = {
      currentTarget: { blur },
      nativeEvent: { x: 0 },
    };
    expect(subject.simulate('click', eventMock));
    expect(clickHandler).not.toHaveBeenCalled();
    expect(blur).not.toHaveBeenCalled();
  });

  describe('with onClick', () => {
    setProp('onClick', () => clickHandler);

    it('should call onClick', () => {
      const blur = jest.fn();
      const eventMock = {
        currentTarget: { blur },
        nativeEvent: { x: 1 },
      };
      expect(subject.simulate('click', eventMock));
      expect(clickHandler).toHaveBeenCalled();
    });
  });
});
