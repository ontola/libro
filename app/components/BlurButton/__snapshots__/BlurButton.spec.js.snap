// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlurButton should render 1`] = `
<button
  onClick={[Function]}
  type="button"
/>
`;
